content/blog/2022/2022-08-03-feature-flags-in-nodejs-react-client-with-launchdarkly.md

```diff
@@ -0,0 +1,332 @@
+
+```javascript
+import React, { Component } from 'react';
+import styled from "styled-components";
+import './App.css';
+import * as LDClient from 'launchdarkly-js-client-sdk';
+
+const theme = {
+  blue: {
+    default: "#3f51b5",
+    hover: "#283593"
+  }
+};
+
+const Button = styled.button`
+    background-color: ${(props) => theme[props.theme].default};
+    color: white;
+    padding: 5px 15px;
+    border-radius: 5px;
+    outline: 0;
+    text-transform: uppercase;
+    margin: 10px 0px;
+    cursor: pointer;
+    box-shadow: 0px 2px 2px lightgray;
+    transition: ease background-color 250ms;
+    &:hover {
+      background-color: ${(props) => theme[props.theme].hover};
+    }
+    &:disabled {
+      cursor: default;
+      opacity: 0.7;
+    }
+  `;
+
+const clickMe = () => {
+  alert("You clicked me!");
+};
+
+class App extends Component {
+  constructor() {
+    super()
+    this.state = {}
+  }
+  componentDidMount() {
+    const user = {
+      // UI based user
+      key: 'user_a'
+    }
+    // SDK requires Client-side ID for UI call
+    this.ldclient = LDClient.initialize('62e*********************', user);
+    this.ldclient.on('ready', this.onLaunchDarklyUpdated.bind(this));
+    this.ldclient.on('change', this.onLaunchDarklyUpdated.bind(this));
+  }
+  onLaunchDarklyUpdated() {
+    this.setState({
+      featureFlags: {
+        hideUser: this.ldclient.variation('hidden-user', '')
+      }
+    })
+  }
+  render() {
+    if (!this.state.featureFlags) {
+      return <div className="App">Loading....</div>
+    }
+    return (
+      <div className="App">
+        <div>{ 
+            this.state.featureFlags.hideUser !== 'John Doe'
+              ? <Button theme='blue' onClick={clickMe}>Shiny New Feature</Button>
+              : '' }
+        </div>
+    );
+  }
+}
+
+export default App;
+```
+
+To execute code only when the LaunchDarkly client is ready, we have two mechanisms: an *event* or a *promise*.
+
+With `this.ldclient.on('ready', ...)`, we subscribe to the `ready` event which will fire once the LaunchDarkly client has received the state of all feature flags from the server.
+
+With `this.ldclient.on('change', ...)`, we subscribe to the `change` event which will fire once the LaunchDarkly client has received the state of all feature flags that has got updated or changed in the LaunchDarkly UI.
+
+For the promise mechanism, the SDK supports two methods: `waitUntilReady()` and `waitForInitialization()`. The behavior of `waitUntilReady()` is equivalent to the `ready` event. The promise resolves when the client receives its initial flag data. As with all promises, you can either use `.then()` to provide a callback, or use `await` if you are writing asynchronous code. The other method that returns a promise, `waitForInitialization()`, is similar to `waitUntilReady()` except that it also tells you if initialization fails by rejecting the promise.
```

Comment:
Removed this section as it's not connected to the rest of the article

Reply:
Makes sense..